### ISSUE TITLE
Support hair mods

### ISSUE THREAD

**Issue by focustense** [focustense/easymod#6]
There are a number of mods out that do nothing but replace vanilla NPC hair, including many for the actual vanilla NPCs ([Shiva's replacer](https://www.nexusmods.com/skyrimspecialedition/mods/7922), [Vanilla Hair Replacer](https://www.nexusmods.com/skyrimspecialedition/mods/25574)) but also mod-originating NPCs ([ApachiiHair3DNPC](https://www.nexusmods.com/skyrimspecialedition/mods/14865), [KSHair for CRF](https://www.nexusmods.com/skyrimspecialedition/mods/34804)), and many more.

While the value of creating mugshots for all of these to use as face mods/overrides is highly questionable, there is a use case which is probably closer to what most players actually want, and which may be possible to implement using Nifly and the experienced gained from de-wiggification. The logic would be a **global** setting and look something like:

1. Identify hair mods at startup. Should be mostly auto-detectable, allow users to override.
2. Set up a priority order for identified hair mods based on load order, again allowing this to be overridden.
3. During build, if an NPC is **not** touched by any other face override (i.e. still uses their vanilla face), we can just treat the hair mod as a regular face override, use its headparts and facegen file. This is the easy part.
4. If an NPC **is** modified **but still uses vanilla hair** - i.e. uses a custom face, makeup, etc. but references the same hair-headpart as the original master - **then** use NIF injection to mash-up the winning hair replacer and the selected face.
5. If a modified NPC **does not use** vanilla hair, then _ignore_ the hair replacer - but allow this to be overridden at some level, probably either globally or per-hair-mod.

If we think about the niche that hair replacers tend to fill - providing a quick and easy "overhaul lite" for characters that haven't been fully overhauled - then I believe these steps best reflect the intent of both the mod creators and the modders themselves.

It does need to be stated that this is not as simple as it may sound at first. Two major obstacles I've uncovered are:

1. At least some (maybe most?) hair replacers "cheat" the facegen system by providing _only_ facegens, and using the vanilla names for the `NiNodes`, and don't make _any_ record edits at all. Either they have no ESP, or the ESP is just a dummy plugin to get a BSA to load. While this doesn't make the simple case, where we just copy the facegen file, any more difficult, it makes the NIF injection more complex because it creates ambiguity over the proper headpart names.
    - It's probably (?) safe to assume that they are the originally headparts used by the original master, otherwise they probably (?) wouldn't work in game. However, it's not a sure thing. Without checking each and every one of these files, it's possible that some of them use vanilla headpart names that are not the _original_ vanilla headpart names used by those NPCs. This part of the facegen system is still somewhat of a mystery and many mod creators use trial and error to get through it.
    - If they _are_ in fact using the true original headpart names - i.e. the customized-hair facegens exactly match the headpart list in the master record - then this is relatively simple: just revert any custom hair in the merged plugin back to the vanilla hair, and inject the "not really vanilla" hair from the replacer into the NIF.
    - However, if they are _not_ using the original headparts, just some arbitrary vanilla headparts, then this may become considerably more complicated, as it would be necessary to actually parse the replaced facegens to figure out which nodes are hair/hairline, which is difficult as meshes don't contain any of this metadata.

2. Other replacers, most notably VHR, work by _editing vanilla head parts_, which is an invasive edit (like editing vanilla races) that [requires extensive patching](https://www.nexusmods.com/skyrimspecialedition/articles/1955) to avoid glitches and whose steps can actually break NPC mods that use custom geometry (morphs).
    - We can absolutely do this patching as well, and probably automate the entire process above, without inheriting any of its problems, but the signal for it is different. The app has to actually understand when plugins modify vanilla head parts and inject the modified head parts into any modified NPCs who are still using the original head parts. While this isn't necessarily any more difficult than anything else the app does (especially e.g. de-wigging), it is an entirely different category of functionality that simply doesn't exist and doesn't share much overlap with existing functions.

Bottom line: This is likely to be a highly valuable feature if supported, which can save people many hours or even days of patching time, but it is also going to be one of the most difficult and potentially finicky implementations. Getting this into the beta would be nice, but may get pushed back if it doesn't go smoothly.

**Comment by focustense** [comment]
This is still high priority but I'm going to remove it from the beta milestone and attempt it as a fast-follow. It's a little too ambitious a feature to develop and release without a few dedicated testers.

**Comment by focustense** [comment]
Ref #64 as there is some relation.
